poll_modal: Convert module to TypeScript

pranayjoshi opened the pull request:
<!-- Describe your pull request here.-->

Converts: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**poll_modal.js to  TypeScript**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

pranayjoshi commented:
@zulipbot add "area: typescript migration"

timabbott commented on web/src/poll_modal.ts:
```diff
@@ -21,33 +21,33 @@ function add_option_row(e) {
-        Sortable.create($("#add-poll-form .poll-options-list .simplebar-content")[0], {
-            onUpdate() {},
+        SortableJS.create($("#add-poll-form .poll-options-list .simplebar-content")[0], {
+            // onUpdate() {},
```
What's the story with commenting this line?

pranayjoshi replied:
The onUpdate() {} function threw an "Empty function Exception" when parsed through lint. Therefore, I had to comment it out.

timabbott replied:
Read the linter error more carefully. https://eslint.org/docs/latest/rules/no-empty-function makes clear that adding a comment is a better fix.

pranayjoshi replied:
Hi Tim,
Thanks for pointing it out. I made the required changes.

timabbott commented on web/src/poll_modal.ts:
```diff
@@ -21,33 +21,33 @@ function add_option_row(e) {
     $question_input.on("input", () => {
-        if ($question_input.val().trim() !== "") {
+        const inputValue = $question_input.val();
+        if (inputValue !== null && typeof inputValue === "string" && inputValue.trim() !== "") {
```
You shouldn't need to add if clauses for these values not being strings. `$<HTMLInputElement>` is helpful, but also I think `$question_input.val()!` might be more appropriate here; it'll be a bug that should throw an exception if one of these values is even not a string.

pranayjoshi replied:
Hi, I tried the same approach but the issue is. The `$question_input` can be `string | number | string[]` and trim doesnt exist on any other type except `string`.

pranayjoshi replied:
- There was one solution i.e to use `.toString()` but I thought. `$submit_button.prop("disabled", false);` depends on this, therefore its a bad idea to use it.
- another solution could be to use `.text()` or `.html()` at line 32. But I am not sure if thats what the function's need.

So at the end, checking it using the if else is the best I could come up for without breaking the previous logic.
Although, I am open for any suggestions

timabbott replied:
I expect `.text` would be a reasonable solution. @LaPulgaaa may be able to help with some of these more basic questions.

LaPulgaaa replied:
`$question_input` is attribute of a `input` element.
```
        <input type="text" id="poll-question-input" class="modal_text_input" placeholder="{{t 'Your question'}}" />
```
`.text()` is used to retrieve or set text content of `HTML` element(like `div`, `span`, `p`) not the value attribute of a `form` element like `input`. 
 `.val()` is specifically used for that purpose. I think using `.toString()` might be a better approach.

pranayjoshi replied:
Hi,
I made the requested changes.
Thanks

LaPulgaaa commented on web/src/poll_modal.ts:
```diff
@@ -56,11 +56,13 @@ export function poll_options_setup() {
-    const question = $("#poll-question-input").val().trim();
+export function frame_poll_message_content(): string {
+    const value = $("#poll-question-input").val();
+    const question = typeof value === "string" ? value.trim() : "";
```
[comment](https://github.com/zulip/zulip/pull/29089/commits/20d42e536d9379f54fb56a7b488c3f4b3cb34507#r1506010886)

timabbott commented:
It looks like you've created a merge commit instead of rebasing your branch, which is not recommended. Check out our [GitHub guide](https://zulip.readthedocs.io/en/latest/git-guide.html) and [commit guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for advice on how to set up Git to not create them when using `git pull` and other important advice.

pranayjoshi commented:
> It looks like you've created a merge commit instead of rebasing your branch, which is not recommended. Check out our [GitHub guide](https://zulip.readthedocs.io/en/latest/git-guide.html) and [commit guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for advice on how to set up Git to not create them when using `git pull` and other important advice.
Thanks, for reminding. I made the required changes.

pranayjoshi commented:
Hi any updates??

LaPulgaaa commented:
Your commit structure is still not up-to-the mark. It has two different commits with exact same message.

pranayjoshi commented:
> Your commit structure is still not up-to-the mark. It has two different commits with exact same message. The second commit addresses the review comment made originally for the first commit . There is no need for the second commit **here**..Just update your first commit with changes recommended..Going through the [commit guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) would give a better idea.

does this looks fine?

timabbott commented on web/src/poll_modal.ts:
```diff
@@ -1,15 +1,15 @@
 }
 
-function add_option_row(e) {
+function add_option_row(e: {target: JQuery<HTMLInputElement>}): void {
```
You should type `e` properly, rather than one of its fields; you can use types like `JQuery.ClickEvent`. `git grep "e: J"` will find you examples.

andersk commented on web/src/poll_modal.ts:
```diff
@@ -56,11 +57,11 @@ export function poll_options_setup() {
-export function frame_poll_message_content() {
-    const question = $("#poll-question-input").val().trim();
+export function frame_poll_message_content(): string {
+    const question = $("#poll-question-input").val()!.toString().trim();
```
No, use `$<HTMLInputElement>("input#poll-question-input").val()`.

andersk commented on web/src/poll_modal.ts:
```diff
@@ -56,11 +57,11 @@ export function poll_options_setup() {
     const options = $(".poll-option-input")
         .map(function () {
-            return $(this).val().trim();
+            return $(this).val()?.toString().trim();
```
No, use `$<HTMLInputElement>("input.poll-option-input")` and not `.toString()`.

pranayjoshi replied:
what about line 12? Is it correct?

LaPulgaaa replied:
Nope..I think my idea was wrong.

pranayjoshi replied:
I am talking about line 12, the line that Tim mentioned.

timabbott commented:
@pranayjoshi are you still working on this? If not, maybe @LaPulgaaa do you want to just do a new PR for this file conversion?

pranayjoshi commented:
> @pranayjoshi are you still working on this? If not, maybe @LaPulgaaa do you want to just do a new PR for this file conversion?

Hey sorry I am busy in my coursework. Sure, @LaPulgaaa can work on this

LaPulgaaa commented:
Cool . Thanks @pranayjoshi .

LaPulgaaa commented:
Now that `poll_modal.js` is migrated , I think we close this PR @pranayjoshi .

andersk commented:
- Closed in favor of #29374.
